### ISSUE TITLE
Warning on Corrade

### ISSUE THREAD

**Issue by AndySomogyi** [mosra/corrade#53]
Corrade build on OSX results in this somewhat annoying linker warning. This warning has been around since I first started using Corrade. 

The warning shows up in both Eclipse (Unix Makefile) and Xcode builds. 

I've haven't been bothered enough to try to track it down, but just reporting it here if anyone has a quick fix. 

This warning does not seem to effect functionality or usage, it's just kind of annoying to see the warning on each build. 

```
ld: warning: direct access in function 'Corrade::Utility::ConfigurationGroup::addGroup(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, Corrade::Utility::ConfigurationGroup*)' from file '../../extern/corrade/src/Corrade/Utility/libCorradeUtility.a(ConfigurationGroup.cpp.o)' to global weak symbol 'std::__1::char_traits<char>::eq(char, char)' from file 'CMakeFiles/CylinderTest.dir/AppDelegate.mm.o' means the weak symbol cannot be overridden at runtime. This was likely caused by different translation units being compiled with different visibility settings.
ld: warning: direct access in function 'Corrade::Utility::ConfigurationGroup::setValueInternal(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, unsigned int, Corrade::Containers::EnumSet<Corrade::Utility::ConfigurationValueFlag, (unsigned char)255>)' from file '../../extern/corrade/src/Corrade/Utility/libCorradeUtility.a(ConfigurationGroup.cpp.o)' to global weak symbol 'std::__1::char_traits<char>::eq(char, char)' from file 'CMakeFiles/CylinderTest.dir/AppDelegate.mm.o' means the weak symbol cannot be overridden at runtime. This was likely caused by different translation units being compiled with different visibility settings.
ld: warning: direct access in function 'Corrade::Utility::ConfigurationGroup::addValueInternal(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, Corrade::Containers::EnumSet<Corrade::Utility::ConfigurationValueFlag, (unsigned char)255>)' from file '../../extern/corrade/src/Corrade/Utility/libCorradeUtility.a(ConfigurationGroup.cpp.o)' to global weak symbol 'std::__1::char_traits<char>::eq(char, char)' from file 'CMakeFiles/CylinderTest.dir/AppDelegate.mm.o' means the weak symbol cannot be overridden at runtime. This was likely caused by different translation units being compiled with different visibility settings.
```

**Comment by mosra** [comment]
Hmm. I did't encounter this warning myself -- reminds me of those dreaded warnings about DLL interface for STL classes that MSVC likes to warn about. Looking at the Travis build, [it prints two warnings](https://travis-ci.org/mosra/corrade/jobs/436829707#L200-L203) but these are completely different and unrelated to this. Checked also the iOS build and my local Linux Clang/libc++ build, but no warning there either.

Anything special you have enabled? Debug build or Release? I see this is a static build? Looking at it, it seems to be caused by a particular STL implementation (libc++), where it has some attributes that trigger this. But no idea why just here and not elsewhere.

As a completely random idea, could you try compiling with this patch applied?

```diff
diff --git a/src/Corrade/Utility/VisibilityMacros.h b/src/Corrade/Utility/VisibilityMacros.h
index e741f559..16fb1a73 100644
--- a/src/Corrade/Utility/VisibilityMacros.h
+++ b/src/Corrade/Utility/VisibilityMacros.h
@@ -50,13 +50,11 @@
 #endif
 
 /** @hideinitializer
  * @brief Public symbol in a static library
  */
-#ifdef CORRADE_TARGET_WINDOWS
 #define CORRADE_VISIBILITY_STATIC
-#else
-#define CORRADE_VISIBILITY_STATIC __attribute__ ((visibility ("default")))
-#endif
 
 /** @hideinitializer
 @brief Local symbol
```

**Comment by mosra** [comment]
Hmm. I'm getting a similar warning on Travis when compiling Magnum:

```
ld: warning: direct access in function 'Magnum::Shaders::AbstractVector<2u>::AbstractVector()' from file 'src/Magnum/Shaders/CMakeFiles/MagnumShadersObjects.dir/DistanceFieldVector.cpp.o' to global weak symbol 'vtable for Magnum::Shaders::AbstractVector<2u>' from file 'src/Magnum/Shaders/CMakeFiles/MagnumShadersObjects.dir/AbstractVector.cpp.o' means the weak symbol cannot be overridden at runtime. This was likely caused by different translation units being compiled with different visibility settings.
ld: warning: direct access in function 'Magnum::Shaders::AbstractVector<3u>::AbstractVector()' from file 'src/Magnum/Shaders/CMakeFiles/MagnumShadersObjects.dir/DistanceFieldVector.cpp.o' to global weak symbol 'vtable for Magnum::Shaders::AbstractVector<3u>' from file 'src/Magnum/Shaders/CMakeFiles/MagnumShadersObjects.dir/AbstractVector.cpp.o' means the weak symbol cannot be overridden at runtime. This was likely caused by different translation units being compiled with different visibility settings.
...
```

If you would have a chance testing with the above patch, that would be great :+1: I only have macOS on Travis and doing any trial-and-error testing there is fairly time consuming ;)

**Comment by AndySomogyi** [comment]
I’ll give it a shot tomorrow, I was buried very deep in message pumps, interrupt requests, signal handlers and all that fun stuff last few days, just now looking around.

**Comment by AndySomogyi** [comment]
Hi, 

Just pulled latest from master, and still get a LOT of those visibility warnings:

```

**Comment by mosra** [comment]
Well, I didn't apply [the above patch](https://github.com/mosra/corrade/issues/53#issuecomment-426793169) because I was not sure if it fixes anything for you ;) Could you try applying it locally and building again, to see if it makes any difference? Thanks!

For the `text-based stub file` warnings I have no idea. Not a problem of Corrade, I think.

**Comment by mosra** [comment]
The stable release is out, so we can experiment again :)

I googled around and this popped up: https://github.com/avast-tl/retdec/pull/349/commits/0f90931787ee707a9579827b4869551cc4ab7de8. Could you try building the whole project (maybe Corrade as well) with `-fvisibility-inlines-hidden` added to `CMAKE_CXX_FLAGS`?

**Comment by mosra** [comment]
Should be fixed in 7ae90319e42871efaa5c59395ee5d3e18e388e47.
